jaeger-core/src/main/java/io/jaegertracing/internal/propagation/TraceContextCodec.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2019, The Jaeger Authors
+ * Copyright (c) 2017, Uber Technologies, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.jaegertracing.internal.propagation;
+
+import io.jaegertracing.internal.JaegerObjectFactory;
+import io.jaegertracing.internal.JaegerSpanContext;
+import io.jaegertracing.internal.utils.Utils;
+import io.jaegertracing.spi.Codec;
+import io.opentracing.propagation.TextMap;
+
+import java.util.Collections;
+import java.util.Map;
+
+
+/**
+ * This format follows the Trace Context specification  https://www.w3.org/TR/trace-context/
+ *
+ * <p>
+ * Example usage:
+ *
+ * <pre>{@code
+ * traceContextCodec = new TraceContextCodec();
+ * tracer = new JaegerTracer.Builder(serviceName, reporter, sampler)
+ *                    .registerInjector(Format.Builtin.HTTP_HEADERS, traceContextCodec)
+ *                    .registerExtractor(Format.Builtin.HTTP_HEADERS, traceContextCodec)
+ *                    ...
+ * }</pre>
+ *
+ * <p>
+ */
+public class TraceContextCodec implements Codec<TextMap> {
+  protected static final String TRACE_CONTEXT_NAME = "tracecontext";
+
+  private static final int VERSION_SIZE = 2;
+  private static final int TRACEPARENT_DELIMITER_SIZE = 1;
+  private static final int TRACE_ID_HEX_SIZE = 32;
+  private static final int SPAN_ID_HEX_SIZE = 16;
+  private static final int TRACE_ID_OFFSET = VERSION_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int SPAN_ID_OFFSET =
+      TRACE_ID_OFFSET + TRACE_ID_HEX_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int TRACE_OPTION_OFFSET =
+      SPAN_ID_OFFSET + SPAN_ID_HEX_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int SAMPLED_FLAG = 1;
+
+  private static final byte JAEGER_SAMPLED_FLAG = 1;
+
+
+  private final JaegerObjectFactory objectFactory;
+  private final Long version;
+
+  private TraceContextCodec(Builder builder) {
+    this.objectFactory = builder.objectFactory;
+    this.version = builder.version;
+  }
+
+  @Override
+  public void inject(JaegerSpanContext spanContext, TextMap carrier) {
+    long traceIdHigh = spanContext.getTraceIdHigh();
+
+    //From the specification:
+    //When a system operates with a trace-id that is shorter than 16 bytes,
```

Comment:
this is a terrible idea. sometimes IDs are truncated due to old instrumentation and this will ensure you can't match them later. is this really in the spec?

Reply:
I mean imagine if upstream is 128-bit a middle hop accidentally truncates making several outbound client calls. now you have effectively a new trace for all those client calls as there's no way to detect the truncation anymore, unless you process the entire tree and make some assumptions. All the last years heuristics to match no-longer work.

Reply:
also if you choose against changing the trace ID you were given, you have the second benefit of a consistent subtree. as this ends up in logs it is helpful. If it turns out the intent was not a 64-bit, rather a truncation, you would see that also in the logs.. and see the downstream affected. if everyone changes the trace id upon seeing a short one, not only can you not easily tell, but also nothing matches. I hope this makes sense.

Reply:
I don't disagree that adding this additional information in makes it harder to use in older systems, but the spec does seem pretty rigid about this. 

> Because tracing systems may make sampling decisions based on the value of trace-id, for increased interoperability vendors MUST keep the random part of trace-id ID on the left side.
> 
> When a system operates with a trace-id that is shorter than 16 bytes, it SHOULD fill-in the extra bytes with random values rather than zeroes. Let's say the system works with an 8-byte trace-id like 3ce929d0e0e4736. Instead of setting trace-id value to 0000000000000003ce929d0e0e4736 it SHOULD generate a value like 4bf92f3577b34da6a3ce929d0e0e4736 where 4bf92f3577b34da6a is a random value or a function of time and host value.

Reply:
The spec does not say that you need to correct this. The spec says that if you generate it you should follow that rule. As I mentioned in a different comment the left part can be 0 if it is randomly generated

Reply:
agree with @bogdandrutu's interpretation. 

Having said that, that's what happens when specs are written without reference implementation. What if we all join forces and create a single reusable library for Zipkin, Jaeger, OpenCensus, and OpenTelemetry, for parsing W3C trace context? All 4 projects have identical format of trace IDs. The `tracestate` might need some customization.

Reply:
Going to remove this - thanks for the help clarifying :) 

Curious how we should handle the case where users have configured the tracer to not be 128 bit but register the trace context codec